Send dropping messages over 438 characters

I have an application that uses the Send method to send a json encoded byte array and it seems to be dropping messages when the message data is over 438 bytes. Is there some flag I need to set for longer strings?

I'm using Azure Event Hubs (which shows no messages received when I only send messages over 438 bytes)

```
serializedEvent, err := eh.codec.Encode(eh.index, &event.Content)
  if err != nil {
    if !event.Guaranteed() {
      return false
    }

    logp.Critical("Unable to encode event: %v", err)
    return false
  }

  session, err := eh.client.NewSession()
  if err != nil {
    logp.Critical("Creating AMQP session: %v", err)
    return false
  }

  sender, err := session.NewSender(amqp.LinkTargetAddress("/" + eh.queue))
  if err != nil {
    logp.Critical("Creating sender link: %v", err)
    return false
  }

  //message := amqp.NewMessage(serializedEvent)
  message := &amqp.Message{
    Data: [][]byte{serializedEvent[:438], serializedEvent[438:]},
  }
  logp.Info("Sending message: %v", message)
  err = sender.Send(context.Background(), message)
  defer sender.Close()
  if err != nil {
    logp.Critical("Sending message: %v", err)
    return false
  }
  return true
```

Here's some debug output I've captured when sending a message over 438 bytes.
```
$ go test -v -timeout 100s -run 'TestIssue54_MessagesOver438Chars' -tags "integration debug"
=== RUN   TestIssue54_MessagesOver438Chars
=== RUN   TestIssue54_MessagesOver438Chars/issue54-messages-over-438-chars
TX: Begin{RemoteChannel: 0, NextOutgoingID: 0, IncomingWindow: 5000, OutgoingWindow: 4294967295, HandleMax: 0, OfferedCapabilities: [], DesiredCapabilities: [], Properties: map[]}
RX: Begin{RemoteChannel: 0, NextOutgoingID: 1, IncomingWindow: 5000, OutgoingWindow: 5000, HandleMax: 255, OfferedCapabilities: [], DesiredCapabilities: [], Properties: map[]}
TX: Attach{Name: AGvQHzLfvPCkPdMrAnfyOlkovNrauyoMBDDZgVUM, Handle: 0, Role: Sender, SenderSettleMode: <nil>, ReceiverSettleMode: second, Source: <nil>, Target: source{Address: integration-issue54-11059684649119627575/Partitions/0, Durable: 0, ExpiryPolicy: , Timeout: 0, Dynamic: false, DynamicNodeProperties: map[], Capabilities: []}, Unsettled: map[], IncompleteUnsettled: false, InitialDeliveryCount: 0, MaxMessageSize: 9223372036854775807, OfferedCapabilities: [], DesiredCapabilities: [], Properties: map[]}
RX(Session): Attach{Name: AGvQHzLfvPCkPdMrAnfyOlkovNrauyoMBDDZgVUM, Handle: 0, Role: Receiver, SenderSettleMode: <nil>, ReceiverSettleMode: second, Source: <nil>, Target: source{Address: integration-issue54-11059684649119627575/Partitions/0, Durable: 0, ExpiryPolicy: session-end, Timeout: 0, Dynamic: false, DynamicNodeProperties: map[], Capabilities: []}, Unsettled: map[], IncompleteUnsettled: false, InitialDeliveryCount: 0, MaxMessageSize: 262144, OfferedCapabilities: [], DesiredCapabilities: [], Properties: map[]}
RX(Session): Flow{NextIncomingID: 0, IncomingWindow: 5000, NextOutgoingID: 1, OutgoingWindow: 5000, Handle: 0, DeliveryCount: 0, LinkCredit: 1000, Available: 0, Drain: false, Echo: false, Properties: map[]}
RX(Session): Detach{Handle: 0, Closed: true, Error: *Error{Condition: amqp:not-allowed, Description: , Info: map[]}}
RX: Detach{Handle: 0, Closed: true, Error: *Error{Condition: amqp:not-allowed, Description: , Info: map[]}}
TX(Session): Detach{Handle: 0, Closed: true, Error: *Error(nil)}
--- FAIL: TestIssue54_MessagesOver438Chars (53.89s)
	integration_test.go:681: Creating hub integration-issue54-11059684649119627575
    --- FAIL: TestIssue54_MessagesOver438Chars/issue54-messages-over-438-chars (2.37s)
    	integration_test.go:431: Expected nil, got error link detached, reason: *Error{Condition: amqp:not-allowed, Description: , Info: map[]}
FAIL
exit status 1
FAIL	pack.ag/amqp	53.916s
```

Here's the test I'm running as well:
```go
func TestIssue54_MessagesOver438Chars(t *testing.T) {
	hubName, _, cleanup := newTestHub(t, "issue54")
	defer cleanup()

	label := "issue54-messages-over-438-chars"
	t.Run(label, func(t *testing.T) {
		checkLeaks := leaktest.CheckTimeout(t, 60*time.Second)

		// Create client
		client := newEHClient(t, "issue54")
		defer client.Close()

		// Open a session
		session, err := client.NewSession()
		if err != nil {
			t.Fatal(err)
		}

		senderPartitionPath := hubName + "/Partitions/0"
		// Create a sender
		sender, err := session.NewSender(
			amqp.LinkTargetAddress(senderPartitionPath),
			amqp.LinkReceiverSettle(amqp.ModeSecond),
		)
		if err != nil {
			t.Fatalf("%+v\n", err)
		}

		data := make([]byte, 439)
		rand.Read(data)
		err = sender.Send(context.Background(), amqp.NewMessage(data))
		if err != nil {
			t.Fatal("Expected nil, got error", err)
		}

		var wg sync.WaitGroup
		wg.Add(1)
		var receiveErr error
		go func() {
			defer wg.Done()

			// Create a receiver
			receiverPartitionPath := hubName + "/ConsumerGroups/$Default/Partitions/0"
			receiver, err := session.NewReceiver(
				amqp.LinkSourceAddress(receiverPartitionPath),
				amqp.LinkCredit(10),
				amqp.LinkBatching(false),
			)
			if err != nil {
				receiveErr = err
				return
			}
			defer receiver.Close()
			ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
			msg, err := receiver.Receive(ctx)
			cancel()
			if err != nil {
				receiveErr = fmt.Errorf("Error after receives: %+v", err)
				return
			}

			if !bytes.Equal([]byte(data), msg.GetData()) {
				receiveErr = fmt.Errorf("Expected received message to be %v, but it was %v", string(data), string(msg.GetData()))
			}
		}()

		wg.Wait()
		if receiveErr != nil {
			t.Fatal("Receive error:", receiveErr)
		}

		client.Close() // close before leak check

		checkLeaks()
	})
}
```

--- a/client.go
+++ b/client.go
@@ -597,32 +597,31 @@ func (s *Session) mux(remoteBegin *performBegin) {
 			}
 
 		case fr := <-txTransfer:
+
 			// record current delivery ID
+			var deliveryID uint32
 			if fr.DeliveryID != nil {
-				deliveryIDByHandle[fr.Handle] = *fr.DeliveryID
+				deliveryID = *fr.DeliveryID
+				deliveryIDByHandle[fr.Handle] = deliveryID
 
 				// add to handleByDeliveryID if not sender-settled
 				if !fr.Settled {
-					handlesByDeliveryID[*fr.DeliveryID] = fr.Handle
+					handlesByDeliveryID[deliveryID] = fr.Handle
 				}
+			} else {
+				// if fr.DeliveryID is nil it must have been added
+				// to deliveryIDByHandle already
+				deliveryID = deliveryIDByHandle[fr.Handle]
 			}
 
 			// frame has been sender-settled, remove from map
 			if fr.Settled {
-				deliveryID, ok := deliveryIDByHandle[fr.Handle]
-				if !ok { // this should never happen
-					panic("missing delivery id from deliveryIDByHandle")
-				}
 				delete(handlesByDeliveryID, deliveryID)
 			}
 
 			// if confirmSettlement requested, add done chan to map
 			// and clear from frame so conn doesn't close it.
 			if fr.confirmSettlement && fr.done != nil {
-				deliveryID, ok := deliveryIDByHandle[fr.Handle]
-				if !ok { // this should never happen
-					panic("missing delivery id from deliveryIDByHandle")
-				}
 				settlementByDeliveryID[deliveryID] = fr.done
 				fr.done = nil
 			}

--- a/client.go
+++ b/client.go
@@ -12,6 +12,7 @@ import (
 	"net"
 	"net/url"
 	"sync"
+	"sync/atomic"
 	"time"
 )
 
@@ -166,6 +167,8 @@ type Session struct {
 	allocateHandle   chan *link // link handles are allocated by sending a link on this channel, nil is sent on link.rx once allocated
 	deallocateHandle chan *link // link handles are deallocated by sending a link on this channel
 
+	nextDeliveryID uint32 // atomically accessed sequence for deliveryIDs
+
 	// used for gracefully closing link
 	close     chan struct{}
 	closeOnce sync.Once
@@ -275,6 +278,7 @@ func (s *Sender) Send(ctx context.Context, msg *Message) error {
 		sndSettleMode  = s.link.senderSettleMode
 		rcvSettleMode  = s.link.receiverSettleMode
 		senderSettled  = sndSettleMode != nil && *sndSettleMode == ModeSettled
+		deliveryID     = atomic.AddUint32(&s.link.session.nextDeliveryID, 1)
 	)
 
 	// use uint64 encoded as []byte as deliveryTag
@@ -284,6 +288,7 @@ func (s *Sender) Send(ctx context.Context, msg *Message) error {
 
 	fr := performTransfer{
 		Handle:        s.link.handle,
+		DeliveryID:    &deliveryID,
 		DeliveryTag:   deliveryTag,
 		MessageFormat: &msg.Format,
 		More:          s.buf.len() > 0,
@@ -315,6 +320,7 @@ func (s *Sender) Send(ctx context.Context, msg *Message) error {
 		}
 
 		// clear values that are only required on first message
+		fr.DeliveryID = nil
 		fr.DeliveryTag = nil
 		fr.MessageFormat = nil
 	}
@@ -369,8 +375,8 @@ func (s *Session) mux(remoteBegin *performBegin) {
 		linksByName = make(map[string]*link) // maping of names to links
 		nextHandle  uint32                   // next handle # to be allocated
 
-		handlesByDeliveryID = make(map[uint32]uint32) //mapping of deliveryIDs to handles
-		nextDeliveryID      uint32                    // next deliveryID
+		handlesByDeliveryID = make(map[uint32]uint32) // mapping of deliveryIDs to handles
+		deliveryIDByHandle  = make(map[uint32]uint32) // mapping of handles to latest deliveryID
 
 		settlementByDeliveryID = make(map[uint32]chan deliveryState)
 
@@ -459,6 +465,7 @@ func (s *Session) mux(remoteBegin *performBegin) {
 		// handle deallocation request
 		case l := <-s.deallocateHandle:
 			delete(links, l.remoteHandle)
+			delete(deliveryIDByHandle, l.handle)
 			close(l.rx) // close channel to indicate deallocation
 
 		// incoming frame for link
@@ -590,28 +597,33 @@ func (s *Session) mux(remoteBegin *performBegin) {
 			}
 
 		case fr := <-txTransfer:
-			// nil DeliveryID indicates first message
-			if fr.DeliveryID == nil {
-				// allocate new DeliveryID
-				id := nextDeliveryID
-				nextDeliveryID++
-				fr.DeliveryID = &id
+			// record current delivery ID
+			if fr.DeliveryID != nil {
+				deliveryIDByHandle[fr.Handle] = *fr.DeliveryID
 
 				// add to handleByDeliveryID if not sender-settled
 				if !fr.Settled {
-					handlesByDeliveryID[id] = fr.Handle
+					handlesByDeliveryID[*fr.DeliveryID] = fr.Handle
 				}
 			}
 
 			// frame has been sender-settled, remove from map
 			if fr.Settled {
-				delete(handlesByDeliveryID, *fr.DeliveryID)
+				deliveryID, ok := deliveryIDByHandle[fr.Handle]
+				if !ok { // this should never happen
+					panic("missing delivery id from deliveryIDByHandle")
+				}
+				delete(handlesByDeliveryID, deliveryID)
 			}
 
 			// if confirmSettlement requested, add done chan to map
 			// and clear from frame so conn doesn't close it.
 			if fr.confirmSettlement && fr.done != nil {
-				settlementByDeliveryID[*fr.DeliveryID] = fr.done
+				deliveryID, ok := deliveryIDByHandle[fr.Handle]
+				if !ok { // this should never happen
+					panic("missing delivery id from deliveryIDByHandle")
+				}
+				settlementByDeliveryID[deliveryID] = fr.done
 				fr.done = nil
 			}
 

--- a/integration_test.go
+++ b/integration_test.go
@@ -13,9 +13,11 @@ import (
 	"net"
 	"os"
 	"path/filepath"
+	"regexp"
 	"strconv"
 	"strings"
 	"sync"
+	"sync/atomic"
 	"testing"
 	"time"
 
@@ -413,22 +415,6 @@ func TestIntegration_EventHubs_RoundTrip(t *testing.T) {
 			label: "1 roundtrip, large payload",
 			data:  []string{strings.Repeat("Hello", 1000/len("Hello"))},
 		},
-		// {
-		// 	label: "3 roundtrip, small payload",
-		// 	data: []string{
-		// 		"2Hey there!",
-		// 		"2Hi there!",
-		// 		"2Ho there!",
-		// 	},
-		// },
-		// {
-		// 	label: "1000 roundtrip, small payload",
-		// 	data: repeatStrings(1000,
-		// 		"3Hey there!",
-		// 		"3Hi there!",
-		// 		"3Ho there!",
-		// 	),
-		// },
 	}
 
 	for _, tt := range tests {
@@ -444,7 +430,7 @@ func TestIntegration_EventHubs_RoundTrip(t *testing.T) {
 			}
 
 			// Create receivers before sending to simplify offsets
-			receive := createEventHubReceivers(t, hubName, session)
+			receive := createEventHubReceivers(t, hubName, session, len(tt.data))
 
 			// Create a sender
 			sender, err := session.NewSender(
@@ -465,6 +451,7 @@ func TestIntegration_EventHubs_RoundTrip(t *testing.T) {
 			// Receive from all partitions
 			received, errs := receive()
 
+			// check total number of messages received
 			var total int
 			for _, count := range received {
 				total += count
@@ -478,19 +465,21 @@ func TestIntegration_EventHubs_RoundTrip(t *testing.T) {
 				}
 			}
 
+			// check that data matches
 			for _, data := range tt.data {
 				count, ok := received[data]
 				if !ok {
 					t.Errorf("Expected to receive message %q but didn't", data)
 					continue
 				}
-				if count <= 1 {
-					delete(received, data)
-				} else {
+				if count > 1 {
 					received[data]--
+				} else {
+					delete(received, data)
 				}
 			}
 
+			// report any unexpected messages
 			for msg, count := range received {
 				t.Errorf("Received %d extra messages: %q", count, msg)
 			}
@@ -502,7 +491,8 @@ func TestIntegration_EventHubs_RoundTrip(t *testing.T) {
 	}
 }
 
-func createEventHubReceivers(t testing.TB, hubName string, session *amqp.Session) func() (map[string]int, []error) {
+func createEventHubReceivers(t testing.TB, hubName string, session *amqp.Session, expectedMessages int) func() (map[string]int, []error) {
+	// Create a receivers on both partitions
 	var receivers []*amqp.Receiver
 	for i := 0; i < 2; i++ {
 		receiver, err := session.NewReceiver(
@@ -519,11 +509,15 @@ func createEventHubReceivers(t testing.TB, hubName string, session *amqp.Session
 
 	return func() (map[string]int, []error) {
 		var (
-			received = make(map[string]int)
-			errs     = make([]error, len(receivers))
-			wg       sync.WaitGroup
+			received  = make(map[string]int)
+			errs      = make([]error, len(receivers))
+			wg        sync.WaitGroup
+			remaining = int32(expectedMessages)
 		)
-		// Create a receivers on both partitions
+
+		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+
+		// receive from each partition concurrently
 		for rxIdx, receiver := range receivers {
 			wg.Add(1)
 			go func(rxIdx int, receiver *amqp.Receiver) {
@@ -531,9 +525,7 @@ func createEventHubReceivers(t testing.TB, hubName string, session *amqp.Session
 				defer receiver.Close()
 
 				for i := 0; ; i++ {
-					ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
 					msg, err := receiver.Receive(ctx)
-					cancel()
 					if err != nil {
 						errs[rxIdx] = fmt.Errorf("Error after %d receives: %+v", i, err)
 						return
@@ -543,6 +535,12 @@ func createEventHubReceivers(t testing.TB, hubName string, session *amqp.Session
 					msg.Accept()
 
 					received[string(msg.GetData())]++
+
+					// cancel all receivers once expected messages have been received
+					if atomic.AddInt32(&remaining, -1) < 1 {
+						cancel()
+						return
+					}
 				}
 			}(rxIdx, receiver)
 		}
@@ -553,7 +551,7 @@ func createEventHubReceivers(t testing.TB, hubName string, session *amqp.Session
 }
 
 func TestIssue48_ReceiverModeSecond(t *testing.T) {
-	const azDescription = "The format code '0x68' at frame buffer offset '39' is invalid or unexpected."
+	azDescription := regexp.MustCompile(`The format code '0x68' at frame buffer offset '\d+' is invalid or unexpected`)
 
 	hubName, _, cleanup := newTestHub(t, "issue48")
 	defer cleanup()
@@ -602,7 +600,7 @@ func TestIssue48_ReceiverModeSecond(t *testing.T) {
 		if err == nil {
 			t.Fatal("Expected error, got nil")
 		}
-		if err, ok := err.(*amqp.Error); !ok || err.Description != azDescription {
+		if err, ok := err.(*amqp.Error); !ok || !azDescription.MatchString(err.Description) {
 			t.Fatalf("Unexpected error response: %+v", err)
 		}
 
@@ -644,7 +642,7 @@ func TestIssue48_ReceiverModeSecond(t *testing.T) {
 		if err == nil {
 			t.Fatal("Expected error, got nil")
 		}
-		if err, ok := err.(*amqp.Error); !ok || err.Description != azDescription {
+		if err, ok := err.(*amqp.Error); !ok || !azDescription.MatchString(err.Description) {
 			t.Fatalf("Unexpected error response: %+v", err)
 		}
 

--- a/integration_test.go
+++ b/integration_test.go
@@ -14,6 +14,7 @@ import (
 	"os"
 	"path/filepath"
 	"strconv"
+	"strings"
 	"sync"
 	"testing"
 	"time"
@@ -394,6 +395,161 @@ func TestIntegrationClose(t *testing.T) {
 
 		checkLeaks()
 	})
+}
+
+func TestIntegration_EventHubs_RoundTrip(t *testing.T) {
+	hubName, _, cleanup := newTestHub(t, "receive")
+	defer cleanup()
+
+	tests := []struct {
+		label string
+		data  []string
+	}{
+		{
+			label: "1 roundtrip, small payload",
+			data:  []string{"1Hello there!"},
+		},
+		{
+			label: "1 roundtrip, large payload",
+			data:  []string{strings.Repeat("Hello", 1000/len("Hello"))},
+		},
+		// {
+		// 	label: "3 roundtrip, small payload",
+		// 	data: []string{
+		// 		"2Hey there!",
+		// 		"2Hi there!",
+		// 		"2Ho there!",
+		// 	},
+		// },
+		// {
+		// 	label: "1000 roundtrip, small payload",
+		// 	data: repeatStrings(1000,
+		// 		"3Hey there!",
+		// 		"3Hi there!",
+		// 		"3Ho there!",
+		// 	),
+		// },
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.label, func(t *testing.T) {
+			checkLeaks := leaktest.CheckTimeout(t, 60*time.Second)
+
+			client := newEHClient(t, tt.label)
+
+			// Open a session
+			session, err := client.NewSession()
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			// Create receivers before sending to simplify offsets
+			receive := createEventHubReceivers(t, hubName, session)
+
+			// Create a sender
+			sender, err := session.NewSender(
+				amqp.LinkTargetAddress(hubName),
+			)
+			if err != nil {
+				t.Fatalf("%+v\n", err)
+			}
+
+			// Send messages
+			for i, data := range tt.data {
+				err = sender.Send(context.Background(), amqp.NewMessage([]byte(data)))
+				if err != nil {
+					t.Fatalf("Error after %d sends: %+v", i, err)
+				}
+			}
+
+			// Receive from all partitions
+			received, errs := receive()
+
+			var total int
+			for _, count := range received {
+				total += count
+			}
+			if total != len(tt.data) {
+				t.Errorf("Expected %d messages, got %d", len(tt.data), total)
+				for i, err := range errs {
+					if err != nil {
+						t.Errorf("Partition %d got error: %+v", i, err)
+					}
+				}
+			}
+
+			for _, data := range tt.data {
+				count, ok := received[data]
+				if !ok {
+					t.Errorf("Expected to receive message %q but didn't", data)
+					continue
+				}
+				if count <= 1 {
+					delete(received, data)
+				} else {
+					received[data]--
+				}
+			}
+
+			for msg, count := range received {
+				t.Errorf("Received %d extra messages: %q", count, msg)
+			}
+
+			client.Close() // close before leak check
+
+			checkLeaks()
+		})
+	}
+}
+
+func createEventHubReceivers(t testing.TB, hubName string, session *amqp.Session) func() (map[string]int, []error) {
+	var receivers []*amqp.Receiver
+	for i := 0; i < 2; i++ {
+		receiver, err := session.NewReceiver(
+			amqp.LinkSourceAddress(hubName+"/ConsumerGroups/$default/Partitions/"+strconv.Itoa(i)),
+			amqp.LinkSelectorFilter("amqp.annotation.x-opt-offset > '@latest'"),
+			amqp.LinkCredit(10),
+			amqp.LinkBatching(false),
+		)
+		if err != nil {
+			t.Fatalf("Error creating receiver: %v", err)
+		}
+		receivers = append(receivers, receiver)
+	}
+
+	return func() (map[string]int, []error) {
+		var (
+			received = make(map[string]int)
+			errs     = make([]error, len(receivers))
+			wg       sync.WaitGroup
+		)
+		// Create a receivers on both partitions
+		for rxIdx, receiver := range receivers {
+			wg.Add(1)
+			go func(rxIdx int, receiver *amqp.Receiver) {
+				defer wg.Done()
+				defer receiver.Close()
+
+				for i := 0; ; i++ {
+					ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+					msg, err := receiver.Receive(ctx)
+					cancel()
+					if err != nil {
+						errs[rxIdx] = fmt.Errorf("Error after %d receives: %+v", i, err)
+						return
+					}
+
+					// Accept message
+					msg.Accept()
+
+					received[string(msg.GetData())]++
+				}
+			}(rxIdx, receiver)
+		}
+		wg.Wait()
+
+		return received, errs
+	}
 }
 
 func TestIssue48_ReceiverModeSecond(t *testing.T) {

